File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/store/WaitableMapStore.scala
package com.webank.eggroll.core.deepspeed.store

import com.webank.eggroll.core.util.Logging

import java.util.concurrent.ConcurrentHashMap
import scala.concurrent.duration.Duration

class WaitableMapStore extends Logging {
  private val store = new ConcurrentHashMap[K, V]()
  //  private val promiseMap = new ConcurrentHashMap[K, Promise[V]]()

  def set(key: K, value: V): Unit = {
    logInfo(s"set key: $key, value: $value, store count: ${store.size()}")
    store.put(key, value)
    logInfo(s"set key: $key, value: $value, store count: ${store.size()} done")
    //    Option(promiseMap.remove(key)).foreach(_.success(value))
  }

  def get(key: K, timeout: Duration): Option[V] = {
    val startTime = System.currentTimeMillis()
    while (!store.containsKey(key)) {
      logDebug(s"waiting for key: $key, store count: ${store.size()}")
      val elapsedTime = System.currentTimeMillis() - startTime
      if (elapsedTime > timeout.toMillis) {
        logDebug(s"Timeout after waiting for key: $key for $timeout, store count: ${store.size()}")
        return None
      }
      Thread.sleep(1000)
    }
    Option(store.get(key))
    //    Option(store.get(key)).orElse {
    //      val newPromise = Promise[V]()
    //      val existingPromise = promiseMap.putIfAbsent(key, newPromise)
    //
    //      val promiseToUse = if (existingPromise == null) newPromise else existingPromise
    //      val future = promiseToUse.future
    //      Try(Await.result(future, timeout)).toOption
    //    }
  }

  private def longToV(x: Long): V = {
    val buffer = java.nio.ByteBuffer.allocate(java.lang.Long.BYTES)
    buffer.putLong(x)
    buffer.array().toVector
  }

  private def vToLong(x: V): Long = {
    val buffer = java.nio.ByteBuffer.allocate(java.lang.Long.BYTES)
    buffer.put(x.toArray)
    buffer.flip()
    buffer.getLong
  }

  def add(key: K, amount: Long): Long = {
    logInfo(s"add key: $key, amount: $amount, store count: ${store.size()}")
    vToLong(store.compute(key, (_, v) => {
      if (v == null) {
        longToV(amount)
      } else {
        longToV(vToLong(v) + amount)
      }
    }))
  }

  def destroy(): Unit = {
    store.clear()
    //    promiseMap.clear()
  }
}

File: jvm/new_core/src/main/java/com/eggroll/core/deepspeed/store/WaitableMapStore.java
package com.eggroll.core.deepspeed.store;

import lombok.Data;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.concurrent.ConcurrentHashMap;

@Data
public class WaitableMapStore {
    Logger log = LoggerFactory.getLogger(WaitableMapStore.class);

    private ConcurrentHashMap<String, byte[]> store = new ConcurrentHashMap<>();

    public void set(String key, byte[] value) {
        log.info("set key: " + key + ", value: " + Arrays.toString(value) + ", store count: " + store.size());
        store.put(key, value);
        log.info("set key: " + key + ", value: " + Arrays.toString(value) + ", store count: " + store.size() + " done");
    }

    public byte[] get(String key, long timeout) throws InterruptedException {
        long startTime = System.currentTimeMillis();
        while (!store.containsKey(key)) {
            log.info("waiting for key: {}, timeout {} store count: {}",key,timeout, store.size());
            long elapsedTime = (System.currentTimeMillis() - startTime)/1000;
            if (elapsedTime > timeout) {
                log.error("Timeout after waiting for key: " + key + " for " + timeout + ", store count: " + store.size());
                return null;
            }
            Thread.sleep(1000);
        }
        return store.get(key);
    }

    private byte[] longToV(long x) {
        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
        buffer.putLong(x);
        return buffer.array();
    }

    private long vToLong(byte[] bytes) {
        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
        buffer.put(bytes);
        buffer.flip();
        return buffer.getLong();
    }

    public long add(String key, long amount) {

        byte[] compute = store.compute(key, (k, v) -> {
            if (v == null) {
                return longToV(amount);
            } else {
                return longToV(vToLong(v) + amount);
            }
        });
        long result = vToLong(compute);
        log.info("add key: " + key + ", amount: " + amount + ",return " + result + " store count: " + store.size());
        return result;
    }

    public void destroy() {
        store.clear();
    }

    public  static void main(String[] args){
        System.err.println(992/100);
    }
}